### PR TITLE
Feature/optionally write final depth as scalar

### PR DESCRIPTION
**Pull request by HeartofTheForce**
**What issue does this PR address?**
#1351

**Does this PR introduce a breaking change?**
Probably not. Some unit tests were passing when they possibly shouldn't be.

**Please check if the PR fulfills these requirements**
- [x] The commit follows our [guidelines](https://github.com/serilog/serilog/blob/dev/CONTRIBUTING.md)
- [x] Unit Tests for the changes have been added (for bug fixes / features)

**Other information**:
Unsure about the Destructuring enum being changed to a flag, but found what seems like bugs because of it specifically:
- `ValueTuplesAreRecognizedWhenDestructuring` was acting despite destrucuturing set implicitly to false
- `AllTupleLengthsUpToSevenAreSupportedForCapturing` also acting despite destrucuturing being implicitly false
- `MaximumStringLengthNOTEffectiveForString` ~~passing despite being give a 4 length string and a 3 length limit~~ According to method description this should only apply during destructuring

### CODE REVIEW

File: src/Serilog/Configuration/LoggerDestructuringConfiguration.cs [hunk 1]
```diff
@@ -147,7 +149,23 @@ public LoggerConfiguration ByTransforming<TValue>(Func<TValue, object> transform
         public LoggerConfiguration ToMaximumDepth(int maximumDestructuringDepth)
         {
             if (maximumDestructuringDepth < 0) throw new ArgumentOutOfRangeException(nameof(maximumDestructuringDepth));
-            _setMaximumDepth(maximumDestructuringDepth);
+            _setMaximumDepth(maximumDestructuringDepth, false);
+            return _loggerConfiguration;
+        }
+
+        /// <summary>
+        /// When destructuring objects, depth will be limited to 5 property traversals deep to
```
Comment (rob-earwaker) [inline src/Serilog/Configuration/LoggerDestructuringConfiguration.cs]: The default depth is actually 10. See updated summary for the [`ToMaximumDepth(int maximumDestructuringDepth)`](https://github.com/serilog/serilog/blob/dev/src/Serilog/Configuration/LoggerDestructuringConfiguration.cs#L139) overload on dev.

Comment (HeartofTheForce) [reply src/Serilog/Configuration/LoggerDestructuringConfiguration.cs]: I've updated the summary to match the latest version

**Comment by nblumhardt** [comment]
Thanks for the 👍 

This one will take some time to work through; I'll try to carve out some time for it ASAP, but am on vacation for the next two weeks and so only online intermittently.

The tests that failed after the changes were, I think, by design - Serilog always disassembles tuples whether destructuring is enabled or not, for example. AFAIK there shouldn't be any breaking changes required for this design (with some careful overloading of `ToMaximumDepth()`), but if you think you have found some bugs we should raise those separately.

It would also be great if we can avoid changing `Destructuring` - I think the changes there might be an interesting design space to explore at some point, but keeping this PR as narrow and tactical as possible will help to get the changes through.

HTH!

Nick

**Comment by nblumhardt** [comment]
Hi @HeartofTheForce - hope this finds you well! Did you have a chance to consider my comment above, and are you planning to take this one further? Cheers!

**Comment by HeartofTheForce** [comment]
I'm happy to strip out the `Destructuring` changes, only reasons it was added was because I couldn't find a clean way to reuse `PropertyValueConverter.CreatePropertyValue` to only write the final `Scalar` values while avoiding destructuring any types that would result in the maximum depth being exceeded.

Would you prefer having `PropertyValueConverter.CreatePropertyValue`'s method signature changed to force all values to `Scalar` for the final depth?

I can also move the `Destructuring` enum changes to another PR so it can be reviewed/considered separately and then move from there/

**Comment by nblumhardt** [comment]
Thanks for the update! I don't think a separate PR for the destructuring changes would be accepted at this stage, since the role of the type is to publicly represent the syntax of message templates, and making these changes would cause us to drift away from that.

Is there any way we could get the `_writeFinalDepthAsScalar` flag in as a constructor argument to `DepthLimiter`, and implement the behavior there? https://github.com/serilog/serilog/blob/dev/src/Serilog/Capturing/DepthLimiter.cs#L45

**Pull request closed without merging**